src/auth.ts

```diff
@@ -3,32 +3,59 @@ import * as os from 'os';
 import * as path from 'path';
 import * as core from '@actions/core';
 import * as io from '@actions/io';
+import * as exec from '@actions/exec';
 
 export const M2_DIR = '.m2';
 export const SETTINGS_FILE = 'settings.xml';
+export const PRIVATE_KEY_DIR = '.keys';
+export const PRIVATE_KEY_FILE = 'private-key.asc';
 
 export const DEFAULT_ID = 'github';
 export const DEFAULT_USERNAME = 'GITHUB_ACTOR';
 export const DEFAULT_PASSWORD = 'GITHUB_TOKEN';
+export const DEFAULT_GPG_PRIVATE_KEY = '';
+export const DEFAULT_GPG_PASSPHRASE = 'GPG_PASSPHRASE';
 
 export async function configAuthentication(
   id = DEFAULT_ID,
   username = DEFAULT_USERNAME,
-  password = DEFAULT_PASSWORD
+  password = DEFAULT_PASSWORD,
+  gpgPrivateKey = DEFAULT_GPG_PRIVATE_KEY,
+  gpgPassphrase = DEFAULT_GPG_PASSPHRASE
 ) {
   console.log(
     `creating ${SETTINGS_FILE} with server-id: ${id};`,
-    `environment variables: username=\$${username} and password=\$${password}`
+    'environment variables:',
+    `username=\$${username},`,
+    `password=\$${password},`,
+    `and gpg-passphrase=\$${gpgPassphrase}`
   );
   // when an alternate m2 location is specified use only that location (no .m2 directory)
   // otherwise use the home/.m2/ path
-  const directory: string = path.join(
+  const settingsDirectory: string = path.join(
     core.getInput('settings-path') || os.homedir(),
     core.getInput('settings-path') ? '' : M2_DIR
   );
-  await io.mkdirP(directory);
-  core.debug(`created directory ${directory}`);
-  await write(directory, generate(id, username, password));
+  await io.mkdirP(settingsDirectory);
+  core.debug(`created directory ${settingsDirectory}`);
+  await write(
+    settingsDirectory,
+    SETTINGS_FILE,
+    generate(id, username, password, gpgPassphrase)
+  );
+
+  if (gpgPrivateKey !== DEFAULT_GPG_PRIVATE_KEY) {
+    console.log('importing gpg key');
+    const privateKeyDirectory: string = path.join(
+      os.homedir(),
```

Comment:
I know this is the common default but in a shared runner I think we wanted to allow for paths outside the home directory such that I could drop this into a temp dir.

Reply:
Yeah, I was worried about that. While the private key is being stored in a "temporary" directory inside the home directory, two setup-java actions running on a shared runner could potentially conflict.

Are you suggesting that we allow users to specify a a temporary private key directory, similar to what is already done with the `settings-path` option? Or would you rather see setup-java create a temporary directory using [`os.tmpdir()`](https://nodejs.org/api/os.html#os_os_tmpdir)  based on the [repository + unique run ID](https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables) without any prompting. It looks like using `$GITHUB_WORKSPACE/.tmp` or something is an alternative as well.

I was thinking more along the lines of the latter -- creating a known good temporary directory for the user without asking -- since a user shouldn't care about where the private key file is being written because it's being deleted.

Side note -- this comment about shared runners also brings up the potential need for a temporary key ring. After import, the GPG keyring sticks around and can be used by other runs. This wouldn't be a problem with the hosted runner because it's a new environment every time. A better approach for the shared, self-hosted runners might be to [import using a temporary key ring](https://serverfault.com/a/696205) and then clean it up in a [post step](https://help.github.com/en/actions/building-actions/metadata-syntax-for-github-actions#post).

Let me know your thoughts and I can take another crack at this 😄

Reply:
For a temporary directory, you can use `RUNNER_TEMP`:

>The path of the temporary directory for the runner. This directory is guaranteed to be empty at the start of each job, even on self-hosted runners.
https://help.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#runner-context

Code example: https://github.com/actions/toolkit/blob/83dd3ef0f1e5bc93c5ab7072e1edf1715a01ba9d/packages/tool-cache/src/tool-cache.ts#L555

Reply:
Ah, thanks! I see now that `installer.ts` is already using this directory as well. Updated PR to use the temp directory but some more improvements can be made since `installer.ts` apparently already has some logic built in for fallbacks. I'll move that common temp directory logic to a separate file that both `auth.ts` and `installer.ts` use in a future commit.

- - -

The temporary key ring thing is unfortunately more complicated than I originally thought. You can't just create and use a temporary key ring without also telling the maven gpg plugin to use the same configuration.

If we want to isolate the gpg keyring for shared runners, the cleanest way I've found is to specify a new GPG home directory that lives in the temp directory. This means that each job will get its own brand new GPG configuration that is cleaned up before the next job run.

To start, we will have to import the private key via:
```
gpg --homedir $RUNNER_TEMP/.gnupg/ --import --batch private.asc
```

However, using `--homedir` only affects that particular import command. Future GPG commands will use the default GPG home directory which is a problem for us when it comes invoking GPG later with Maven. To fix this, we have to provide a home directory override for the GPG plugin too.

The resulting `settings.xml` file would then look like:
```xml
<servers>
    <server>
      <id>github</id>
      <username>${env.GITHUB_ACTOR}</username>
      <password>${env.GITHUB_TOKEN}</password>
    </server>
    <server>
      <id>gpg.passphrase</id>
      <passphrase>${env.GPG_PASSPHRASE}</passphrase>
    </server>
    <profiles>
        <profile>
            <activation>
                <activeByDefault>true</activeByDefault>
            </activation>
            <properties>
                <gpg.homedir>${env.RUNNER_TEMP}</gpg.homedir>
            </properties>
        </profile>
    </profiles>
</servers>
```

Creating this default GPG home directory override in the Maven settings **will** break existing setup-java users that have scripted their own GPG import solution. We need to generate a different `settings.xml` for users that haven't specified a private key:

```xml
<servers>
    <server>
      <id>github</id>
      <username>${env.GITHUB_ACTOR}</username>
      <password>${env.GITHUB_TOKEN}</password>
    </server>
</servers>
```
_**NOTE:** The second `gpg.passphrase` server section can stay as-is without any harm but we might as well remove it if we're already conditionally generating settings.xml based on parameters._

This is going to be a bit more difficult to implement and it's already late here. I'll get cracking on this when I can.